File: LeetCode/Hard/0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> pq = new PriorityQueue<>((o1, o2) -> o1.val - o2.val);

        for (ListNode l : lists) {
            if (l != null) {
                pq.offer(l);
            }
        }

        ListNode node = new ListNode();
        ListNode curr = node;

        while (!pq.isEmpty()) {
            curr.next = pq.poll();
            curr = curr.next;

            if (curr.next != null) {
                pq.offer(curr.next);
            }
        }

        return node.next;
    }
}

File: LeetCode/Hard/0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun mergeKLists(lists: Array<ListNode?>): ListNode? {
        val pq = PriorityQueue<ListNode>(compareBy { it.`val` })

        for (l in lists) {
            if (l != null) {
                pq.offer(l)
            }
        }

        val node = ListNode()
        var curr = node

        while (pq.isNotEmpty()) {
            curr.next = pq.poll()
            curr = curr.next

            if (curr.next != null) {
                pq.offer(curr.next)
            }
        }

        return node.next
    }
}